compiler/qsc_qasm3/src/angle.rs

```diff
@@ -0,0 +1,435 @@
+        Angle {
+            value: (self.value + other.value) % (1u64 << self.size),
+            size: self.size,
+        }
+    }
+}
+
+impl Sub for Angle {
+    type Output = Self;
+
+    fn sub(self, other: Self) -> Self {
+        assert_eq!(self.size, other.size, "Sizes must be the same");
+        Angle {
+            value: (self.value + (1u64 << self.size) - other.value) % (1u64 << self.size),
+            size: self.size,
+        }
+    }
+}
+
+impl Mul<u64> for Angle {
+    type Output = Self;
+
+    fn mul(self, factor: u64) -> Self {
+        Angle {
+            value: (self.value * factor) % (1u64 << self.size),
+            size: self.size,
+        }
+    }
+}
+
+impl Mul<u128> for Angle {
+    type Output = Self;
+
+    fn mul(self, factor: u128) -> Self {
+        let r = BigInt::from(self.value) * BigInt::from(factor);
+        let r = r % BigInt::from(1u128 << self.size);
+        Angle {
+            value: r.try_into().expect("Value is too large"),
+            size: self.size,
+        }
+    }
+}
+
+impl Div<u64> for Angle {
+    type Output = Self;
+
+    fn div(self, divisor: u64) -> Self {
+        Angle {
+            value: self.value / divisor,
+            size: self.size,
+        }
+    }
+}
+
+impl Div for Angle {
+    type Output = u64;
+
+    fn div(self, other: Self) -> u64 {
+        assert_eq!(self.size, other.size, "Sizes must be the same");
+        self.value / other.value
+    }
+}
+
+impl Neg for Angle {
+    type Output = Self;
+
+    fn neg(self) -> Self {
+        Angle {
+            value: (1u64 << self.size) - self.value,
+            size: self.size,
+        }
+    }
+}
+
+impl AddAssign for Angle {
+    fn add_assign(&mut self, other: Self) {
+        *self = *self + other;
+    }
+}
+
+impl SubAssign for Angle {
+    fn sub_assign(&mut self, other: Self) {
+        *self = *self - other;
+    }
+}
+
+impl MulAssign<u64> for Angle {
+    fn mul_assign(&mut self, factor: u64) {
+        *self = *self * factor;
+    }
+}
+
+impl DivAssign<u64> for Angle {
+    fn div_assign(&mut self, divisor: u64) {
+        *self = *self / divisor;
+    }
+}
+
+impl TryInto<f64> for Angle {
+    type Error = &'static str;
+
+    fn try_into(self) -> Result<f64, Self::Error> {
+        if self.size > 64 {
+            return Err("Size exceeds 64 bits");
+        }
+        let Some(denom) = safe_u64_to_f64(1u64 << self.size) else {
+            return Err("Denominator is too large");
+        };
+        let Some(value) = safe_u64_to_f64(self.value) else {
+            return Err("Value is too large");
+        };
+        let factor = (2.0 * PI) / denom;
+        Ok(value * factor)
+    }
+}
+
+impl From<Angle> for bool {
+    fn from(val: Angle) -> Self {
+        val.value != 0
+    }
+}
+
+impl fmt::Display for Angle {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", (*self).try_into().unwrap_or(f64::NAN))
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
nitpick: while it increases the folder/file count, it would be consistent with the patterns in the rest of the repo to pull these out into a test module in `angle/test.rs`.